src/main/scala/sample/stream/GroupLogFile.scala

```diff
@@ -29,13 +30,18 @@ object GroupLogFile {
         case other                  => "OTHER"
       }.
       // write lines of each group to a separate file
-      foreach {
+      map {
         case (level, groupFlow) =>
           val output = new PrintWriter(new FileOutputStream(s"target/log-$level.txt"), true)
           // close resource when the group stream is completed
           // foreach returns a future that we can key the close() off of
-          groupFlow.foreach(line => output.println(line)).onComplete(_ => Try(output.close()))
+          groupFlow.
+            foreach(line => output.println(line)).
+            andThen { case t => output.close(); t.get }
       }.
+      buffer(maximumLogLevelCount, OverflowStrategy.error).
+      mapAsync(identity).
+      foreach { _ => Unit }.
```

Comment:
I see what you wanted here, but there is a special sink, Sink.onComplete, that basically does the same as L44-45 and considered idiomatic (also more efficient).